### PR TITLE
Add DOPD advantage-gap routing to SDFT (experimental)

### PR DESCRIPTION
**Pull request by ucalyptus**
## Summary

This PR adds DOPD (Dual On-Policy Distillation, https://huggingface.co/papers/2606.30626) support directly to the
existing, already-merged `SDFTTrainer` / `SDFTConfig` (`trl/experimental/sdft/`), instead of introducing a new
`OPSDTrainer` class.

### Why the pivot

The original approach for this PR vendored code from an unmerged sibling PR (#5990) as a starting point for a new
`OPSDTrainer`, then planned to extend that trainer with DOPD's advantage-gap token routing. After a multi-model
design discussion and code review, we concluded that OPSD is not a structurally distinct trainer from SDFT — it's a
config-variant of it. Both share the same core data contract (`prompt` + `privileged_context`, teacher/student
logits over a completion) and the same training loop shape (sample on-policy from the student, score a privileged
teacher, take a divergence-based loss between them). The only real differences are in the *loss function's shape*
(top-k support side, whether the divergence is full-vocab or truncated, how regimes are routed) and a handful of
config defaults — not in the data flow, the optimizer step, or anything else that TRL's actual trainer-splitting
precedent is organized around. TRL splits trainers (DPO vs KTO vs ORPO, GRPO vs online DPO, etc.) along differences
in training loop shape or data contract, not along differences in hyperparameter defaults for what is otherwise the
same loop. Folding DOPD into SDFT keeps one trainer, one config, one set of tests, and one entry point for anyone
doing on-policy self-distillation with a privileged teacher, with DOPD as one more `distillation_mode` choice
alongside `sampled_token`, `full_logits`, and `topk_logits`.

As a side effect, this also makes the PR fully independent of PR #5990 — it no longer vendors or depends on any
code from that unmerged branch.

## What changed

**`SDFTConfig` (`trl/experimental/sdft/sdft_config.py`)** — new fields:
- `distillation_mode` gains a fourth option, `"dopd"`, alongside the existing `"sampled_token"`, `"full_logits"`,
  and `"topk_logits"`.
- `distillation_topk_support: Literal["student", "teacher"]` (default `"student"`) — controls which side's logits
  define the top-k token support for `distillation_mode="topk_logits"`.
- `distillation_kl_clip: float | None` — optional per-token upper bound on the summed vocabulary divergence,
  capping outlier tokens for `full_logits`/`topk_logits` modes.
- `teacher_chat_template_kwargs: dict[str, Any] | None` — kwargs forwarded to the teacher's chat template
  application, independent of the student's.
- Five `distillation_dopd_*` fields controlling DOPD's four-regime advantage-gap router:
  `distillation_dopd_gap_threshold`, `distillation_dopd_confidence_threshold`, `distillation_dopd_light_topk`,
  `distillation_dopd_self_reg_weight`, `distillation_dopd_student_consistency_weight`.

**Loss utilities (`trl/experimental/sdft/loss_utils.py`)**:
- `compute_topk_self_distillation_loss` gained a `topk_support: str = "student"` parameter (previously always
  used the student's top-k to define the support token set; now `"teacher"` is also selectable — DOPD's routing
  uses the teacher's top-k for its light-signal regime, matching the paper's convention).
- `compute_divergence` gained an optional `kl_clip` parameter.
- New `compute_dopd_routed_loss` function: implements the four-regime advantage-gap token router described in the
  DOPD paper (light top-k reverse-KL toward the teacher for low-gap/confident tokens, top-k reverse-KL against a
  stop-gradient student anchor for low-gap/unsure tokens, full-vocabulary JSD for high-gap/teacher-confident
  tokens, and light top-k reverse-KL against a stop-gradient student anchor for high-gap/student-confident
  tokens; ambiguous high-gap/low-confidence tokens fall back to the stop-gradient student anchor regime).

**Trainer dispatch (`trl/experimental/sdft/sdft_trainer.py`)** — routes to `compute_dopd_routed_loss` when
`distillation_mode="dopd"`, alongside a few infra fixes ported in from the original OPSD prototype during this
pivot (see the trainer-side worker's diff for full detail; summarized here at a high level since it landed in
parallel with this PR description):
- Untruncated teacher tokenization, so the privileged-context teacher pass isn't silently clipped relative to the
  student's completion length.
- A guard against the unsupported ZeRO-3 + PEFT combination for the teacher/student adapter setup.
- Pre-registration of the EMA teacher adapter before training starts, rather than lazily on first use.

## Removed

- The entire `trl/experimental/opsd/` directory (`__init__.py`, `loss_utils.py`, `opsd.py`, `opsd_config.py`,
  `opsd_trainer.py`) — this was vendored from PR #5990 as a scaffold for the now-abandoned `OPSDTrainer` approach
  and is no longer needed.

## Tests

- `tests/experimental/test_dopd_routing.py`: repointed its imports from the deleted
  `trl.experimental.opsd.loss_utils` to `trl.experimental.sdft.loss_utils` (all the functions it exercises —
  `compute_divergence`, `compute_dopd_routed_loss`, `compute_full_logit_self_distillation_loss`,
  `compute_sampled_token_self_distillation_loss`, `compute_topk_self_distillation_loss` — now live there). Its
  regime-routing coverage (mutual exclusivity/exhaustiveness, per-regime gradient flow, threshold sensitivity) is
  unchanged.
- Added `TestTopkSupportParameter` to the same file: constructs student/teacher logits with deliberately
  non-overlapping, non-mirror-symmetric top-k token sets and asserts that `topk_support="student"` vs
  `topk_support="teacher"` produce measurably different loss values (proving the parameter actually changes which
  tokens are compared, not just that it runs without error), plus a check that an invalid `topk_support` value
  raises `ValueError`.

## Docs

- `docs/source/paper_index.md`: rewrote the "DOPD: Dual On-Policy Distillation" section's code example to use
  `from trl.experimental.sdft import SDFTConfig, SDFTTrainer` with `distillation_mode="dopd"` plus the five
  `distillation_dopd_*` fields, instead of the old `OPSDConfig`/`OPSDTrainer` example. The prose explanation of
  what DOPD routing does is unchanged (still accurate). The "for more details" link now points at the
  `sdft_trainer` doc page instead of the now-nonexistent `opsd_trainer` page. Also removed a leftover line
  claiming automatic gsm8k `question`/`answer` column mapping — that convenience mapping lived only in the
  deleted `opsd.py` CLI script, not in `SDFTTrainer` itself, so the doc would have overpromised.
- Confirmed `docs/source/_toctree.yml` has no OPSD entry to remove (this branch never added one; PR #5990, which
  would add one, is the separate unmerged PR this branch no longer depends on).

## Test plan

- [x] `ruff check` / `ruff format --check` clean on all touched files.
- [x] `pytest tests/experimental/test_dopd_routing.py -v` — 7 passed (5 existing regime-routing tests + 2 new
      `topk_support` tests), run against the real installed package (not a stub).
- [x] `from trl.experimental.sdft import SDFTConfig, SDFTTrainer` imports cleanly with all trainer-side changes in
      place.
- [x] `pytest tests/experimental/test_self_distillation_trainer_behavior.py` — 15/19 pass; the same 4 fail
      identically on the pre-pivot `main` branch in this sandbox (no GPU/bf16 support here), confirming no
      regression from this PR's changes.
- [ ] Real GPU/DeepSpeed run exercising `distillation_mode="dopd"` end-to-end, and the ZeRO-3+PEFT guard /
      EMA-adapter-pre-registration paths under actual distributed training — not possible in this sandbox (no
      GPU/multi-node available). These infra fixes are ported faithfully from a working reference implementation
      but have only been verified by import/logic review here, not executed under real distributed training.

---

> [!NOTE]
> **Medium Risk**
> Changes the training loss path and distributed metric aggregation for SDFT/SDPO; DOPD adds an extra no-grad privileged-student forward per step, but validation blocks incompatible teacher/server modes.
> 
> **Overview**
> Adds **DOPD (Dual On-Policy Distillation)** as `distillation_mode="dopd"` on [`SDFTConfig`] / [`SDFTTrainer`], routing each completion token through four advantage-gap regimes (privileged teacher vs privileged student for routing; bare student for the loss). New `distillation_dopd_*` hyperparameters control gap/confidence thresholds and regime weights; **`teacher_model_kind="live"`** and the vLLM teacher server are rejected for DOPD because routing needs full local logits and a distinct privileged-student forward.
> 
> Extends shared distillation utilities with **`compute_dopd_routed_loss`**, optional **`distillation_kl_clip`** (sum-then-clamp on vocab divergence), and **`distillation_topk_support`** (`"student"` vs `"teacher"`) for top-k objectives. Docs in `paper_index.md` and `sdft_trainer.md` are updated to describe DOPD and server limitations.
> 
> **SDFT/SDPO trainers** pick up related hardening: untruncated teacher (and teacher-conditioned generation) tokenization with separate **`teacher_chat_template_kwargs`**, QLoRA adapter bf16 cast, EMA teacher adapter registration before DeepSpeed wrap, guards for ZeRO-3+PEFT, Liger+PEFT/logit-scale incompatibilities, sequence-parallelism rejection, and **`_global_masked_mean`** for distributed distillation metrics. CPU tests in `test_dopd_routing.py` cover regime routing, gradients, and config validation.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 79e5e715ba490d5bc43b83486b61bb77248254b2. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by ucalyptus** [comment]
hey @kashif , a good time to open up this conversation?

**Comment by ucalyptus** [comment]
@qgallouedec @kashif is this up for review or we can park for later?

**Comment by ucalyptus** [comment]
@qgallouedec review please

**Comment by ucalyptus** [comment]
@kashif how about now?

**Comment by ucalyptus** [comment]
@kashif @qgallouedec lmk if we are moving on this

File: trl/experimental/sdft/loss_utils.py [hunk 1]
```diff
@@ -133,3 +156,101 @@ def compute_sampled_token_self_distillation_loss(
...
+    # detached student anchor. Preserves confident student exploration without pulling toward the unsure
+    # teacher on high-gap tokens.
+    loss4 = student_consistency_weight * compute_topk_self_distillation_loss(
+        student_logits,
+        student_logits.detach(),
+        distillation_topk=light_topk,
+        distillation_alpha=1.0,
+        distillation_add_tail=True,
+        topk_support="student",
+    )
```
[the remark this answers is not in the document]
Comment (ucalyptus) [reply trl/experimental/sdft/loss_utils.py:252]: Verified and fixed — this was a real bug, and the paper confirms it.

Regimes 2 and 4 anchored on the bare student's own detached logits: KL(π ‖ sg(π)) is identically zero in value and gradient, so both regimes were dead loss terms and the two weights had no training effect. Numeric check on the exact top-k + tail + alpha=1.0 path with fixed support indices: loss = 0.0, max |grad| = 0.0 (control against a real teacher: 0.66).

The paper (arXiv 2606.30626, eqs. 7 and 9) defines both the LL and HS regimes as reverse KL of the bare student against a stop-gradient **privileged** student, sg[ΠS(·|x, p, y<n)] — the anchor must be the privileged student, which the trainer already computes under no_grad for routing.

Fixed in 03d5348c:
- loss2/loss4 anchor on `privileged_student_logits.detach()` per eqs. 7/9
- config docstrings name the privileged-student anchor
- test batch now uses a genuinely different privileged student (routing regimes preserved), the exhaustive-mask test mirrors the privileged-based routing masks, and the gradient tests distinguish the degenerate (privileged == bare → zero) case from the live (privileged != bare → nonzero) case

File: trl/experimental/sdft/loss_utils.py [hunk 2]
```diff
@@ -38,6 +39,9 @@ def compute_divergence(
         kl_teacher = F.kl_div(mixture, teacher_log_probs, reduction="none", log_target=True)
         kl_student = F.kl_div(mixture, student_log_probs, reduction="none", log_target=True)
         kl = torch.lerp(kl_student, kl_teacher, alpha)
+    if kl_clip is not None:
+        # Pointwise per-vocabulary-entry clip: caps high-divergence style tokens before the vocabulary sum.
+        kl = kl.clamp(max=kl_clip)
```
[the remark this answers is not in the document]
Comment (kashif) [reply trl/experimental/sdft/loss_utils.py:46]: already fixed in ca0e6e1c ("sdft: clip kl per-token instead of per-vocab-entry"): the vocabulary sum now happens before the clamp (`kl = kl.sum(-1); if kl_clip is not None: kl = kl.clamp(max=kl_clip)`), so the value being clipped is the already-non-negative per-position KL, not individual per-vocab-entry terms. An upper-bound-only clamp on a non-negative value can't go negative. Resolving.

File: trl/experimental/sdft/sdft_config.py [hunk 3]
```diff
@@ -30,14 +30,41 @@ class SDFTConfig(_BaseConfig):
...
+            Distillation objective mode. SDFT defaults to top-k logit distillation. `dopd` routes each token between
+            four regimes based on the teacher/student advantage gap, following DOPD
+            (https://huggingface.co/papers/2606.30626); see the `distillation_dopd_*` parameters.
         distillation_topk (`int`, *optional*, defaults to `100`):
             Number of top tokens used by the default SDFT top-k logit objective.
+        distillation_topk_support (`Literal["student", "teacher"]`, *optional*, defaults to `"student"`):
+            Which side's logits define the top-k token support for `distillation_mode="topk_logits"`. SDFT's
+            convention is `"student"`; some methods built on this trainer (e.g. OPSD) use `"teacher"` instead. Only
+            used when `distillation_mode="topk_logits"`; the `"dopd"` mode's internal top-k regimes always use
+            `"teacher"` support per that method's own convention, regardless of this setting.
```
[the remark this answers is not in the document]
Comment (kashif) [reply trl/experimental/sdft/sdft_config.py]: already fixed in b3588017 ("sdft: floor kl_clip at 0 and drop stale OPSD refs") — no more OPSD references in sdft_config.py. Resolving.

File: trl/experimental/sdft/loss_utils.py [hunk 4]
```diff
@@ -72,19 +76,35 @@ def compute_topk_self_distillation_loss(
...
 ) -> torch.Tensor:
-    """Compute distillation loss on the student's top-k token support.
+    """Compute distillation loss on a top-k token support.
 
-    The student's top-k logits define the support. The teacher distribution is projected onto the same token indices.
-    The selected support is then either renormalized or augmented with a tail bucket before the divergence is computed.
+    `topk_support` selects which side's top-k logits define the support: SDFT's convention is `"student"`; some
+    methods built on this trainer (e.g. OPSD) use `"teacher"` instead. The other side's distribution is projected
+    onto the same token indices. The selected support is then either renormalized or augmented with a tail bucket
+    before the divergence is computed.
```
[the remark this answers is not in the document]
Comment (kashif) [reply trl/experimental/sdft/loss_utils.py]: same as the sdft_config.py thread — fixed in b3588017, no OPSD references left in loss_utils.py. Resolving.

File: trl/experimental/sdft/loss_utils.py [hunk 5]
```diff
@@ -72,19 +76,35 @@ def compute_topk_self_distillation_loss(
     distillation_topk: int,
```
[the remark this answers is not in the document]
Comment (kashif) [reply trl/experimental/sdft/loss_utils.py:78]: ported `kl_clip`/`topk_support` to `trl/experimental/sdpo/loss_utils.py` with the same signatures and body as the sdft copy (defaults preserve SDPO's current behavior exactly; SDPO's own config/trainer don't expose these yet, but the shared helper functions are now aligned so a future fix only needs to land once).

File: trl/experimental/sdft/teacher_sync.py [hunk 6]
```diff
@@ -133,7 +133,11 @@ def _initialize_teacher_adapter(self):
 
         teacher_state = {k: torch.zeros_like(v) for k, v in student_state.items()}
 
-        model.add_adapter(self.teacher_adapter_name, self.teacher_adapter_config)
+        # The teacher adapter may already have been registered by the trainer's __init__ (needed so it
+        # exists before accelerate/DeepSpeed wraps the model under ZeRO-3). Only add it here if that
+        # hasn't happened yet — PEFT rejects re-adding an adapter name that's already registered.
+        if self.teacher_adapter_name not in model.peft_config:
+            model.add_adapter(self.teacher_adapter_name, self.teacher_adapter_config)
```
[the remark this answers is not in the document]
Comment (kashif) [reply trl/experimental/sdft/teacher_sync.py:140]: ported the same guard to `trl/experimental/sdpo/teacher_sync.py`. Note SDPO's trainer doesn't currently pre-register the teacher adapter in `__init__` the way SDFT's does, so this is a no-op today, but it keeps the two copies aligned for when that lands on the SDPO side too.

File: tests/experimental/test_dopd_routing.py [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+        assert not torch.allclose(loss_student_support, loss_teacher_support, atol=1e-4), (
+            "topk_support='student' and topk_support='teacher' produced the same loss despite disjoint top-k "
+            "token support sets; the parameter is not affecting which tokens are compared."
+        )
+
+    def test_invalid_topk_support_raises(self):
+        student_logits, teacher_logits = self._build_disjoint_logits()
+
+        with pytest.raises(ValueError, match="topk_support"):
+            compute_topk_self_distillation_loss(
```
[the remark this answers is not in the document]
Comment (kashif) [reply tests/experimental/test_dopd_routing.py:348]: added `TestKLClip` in tests/experimental/test_dopd_routing.py: one test confirming the clipped value is bounded at the threshold and non-negative on a case with mixed-sign per-vocab-entry KL terms, one confirming kl_clip=None matches the unclipped value.

File: trl/experimental/sdft/sdft_trainer.py [hunk 8]
```diff
@@ -1309,6 +1383,19 @@ def _log_self_distillation_metric(self, mode: str, value: float) -> None:
...
+
+        Averaging each rank's own masked mean with an unweighted `gather(...).mean()` is biased whenever ranks hold
+        different numbers of valid tokens (e.g. variable completion lengths): a rank with few valid tokens would
+        count as much as one with many. Summing the local numerator/denominator and dividing once (GRPO/RLOO's
+        `global_masked_mean` pattern) weights every token equally regardless of which rank it landed on.
+        """
+        local_sum = (values * mask).sum()
+        local_count = mask.sum().float()
+        totals = self.accelerator.reduce(torch.stack([local_sum, local_count]), reduction="sum")
+        return (totals[0] / totals[1].clamp(min=1.0)).item()
```
[the remark this answers is not in the document]
Comment (kashif) [reply trl/experimental/sdft/sdft_trainer.py:1462]: good catch — this is genuinely a bug in my own `_global_masked_mean` addition, not something I checked against bf16 carefully enough. Fixed by casting `local_sum` to float32 before the stack (matches `local_count`'s dtype, and avoids summing many low-precision values as a side benefit) in both trl/experimental/sdft/sdft_trainer.py and trl/experimental/sdpo/sdpo_trainer.py. Pushed in b8266919.

**Comment by ucalyptus** [comment]
@kashif can u approve the workflows?

**Review by kashif — APPROVED** [review]

**Comment by ucalyptus** [comment]
@kashif merge if all good.

**Comment by ucalyptus** [comment]
@kashif lmk if merge-ready
